### PR TITLE
MVCC Storage Node

### PR DESCRIPTION
**Pull request by darkterbear**
- Implemented `MVCCNode`
- For each key, stores a vector of `MVCCVersion`
- Each `MVCCVersion` contains a value, read timestamp (RTS), and write timestamp (WTS)
- Enforces MVCC rules on reading snapshots, and writing new versions

### CODE REVIEW

**Comment by audreyccheng** [comment]
Can you add unit tests in test_storage_handler.hpp?

File: include/txn/base_store.hpp [hunk 1]
```diff
@@ -225,5 +225,187 @@ template <typename K> class LockNode {
...
 };
 
+// template <typename T>
+class MVCCVersion {
+ protected:
+  string value;
+  long wts;
+  long rts;
+  bool is_primary;
+  long visibility;
```
Comment (darkterbear) [inline include/txn/base_store.hpp:235]: Comment what values `visibility` takes on for diff. purposes

File: include/txn/base_store.hpp [hunk 2]
```diff
@@ -225,5 +225,187 @@ template <typename K> class LockNode {
...
+
+  const bool visible_by(const long tts) {
+    return (visibility == tts || visibility == -1) && wts <= tts;
+  }
+
+  const bool globally_visible() {
+    return visibility == -1;
+  }
+
+  void update_rts(const long tts) {
```
Comment (darkterbear) [inline include/txn/base_store.hpp]: long &tts

File: include/txn/base_store.hpp [hunk 3]
```diff
@@ -225,5 +225,187 @@ template <typename K> class LockNode {
...
+   * Gets a relevant snapshot for key k for this transaction.
+   * If original, gets the version that this transaction originally read from (i.e. latest, globally-visible version with wts <= tts).
+   * If !original, gets the latest version visible to this txn (may be a new version in-progress, written by this txn).
+   */
+  MVCCVersion* get_snapshot(const long tts, const K &k, bool original) {
+    if (db.find(k) == db.end()) {
+      return NULL; 
+    }
+
+    vector<MVCCVersion> versions = db.at(k);
```
Comment (darkterbear) [inline include/txn/base_store.hpp]: This makes a copy ;-;

File: include/txn/base_store.hpp [hunk 4]
```diff
@@ -225,5 +225,187 @@ template <typename K> class LockNode {
...
+
+      MVCCVersion *latest_visible_snapshot = get_snapshot(tts, k, false);
+      if (!latest_visible_snapshot->globally_visible()) {
+        // If this txn already created a new version, just modify that version directly (it's visible only to us anyways)
+        latest_visible_snapshot->assign(v);
+        latest_visible_snapshot->assign_primary(is_primary);
+      } else {
+        // ... otherwise, create a new version
+        vector<MVCCVersion> &versions = db.at(k);
+        versions.insert(versions.begin(), MVCCVersion(tts, v, is_primary));
```
Comment (darkterbear) [inline include/txn/base_store.hpp]: just use db.at(k)

**Comment by audreyccheng** [comment]
Looks great!

**Pull request merged**